src/render/painter.js

```diff
@@ -261,6 +274,87 @@ class Painter {
 
         this.frameHistory.record(Date.now(), this.transform.zoom, style.getTransition().duration);
 
+        const layerIds = this.style._order;
+
+        // 3D pass
+        // We first bind an offscreen framebuffer and render each 3D layer to
+        // its own texture, which we'll use later in the translucent pass to
+        // render to the main framebuffer. By doing this before we render to
+        // the main framebuffer we won't have to do an expensive framebuffer
+        // restore mid-render pass.
+        // The most important distinction of the 3D pass is that we use the
+        // depth buffer in an entirely different way (to represent 3D space)
+        // than we do in the 2D pass (to preserve layer order).
+        this.renderPass = '3d';
+        {
+            const gl = this.gl;
+            // We'll wait and only attach the framebuffer if we think we're
+            // rendering something.
+            let fboAttached = false;
+
+            let sourceCache;
+            let coords = [];
+
+            for (let i = 0; i < layerIds.length; i++) {
+                const layer = this.style._layers[layerIds[i]];
+
+                if (!layer.has3DPass()) continue;
+
+                if (layer.paint['fill-extrusion-opacity'] === 0 || layer.isHidden(this.transform.zoom)) {
+                    // Don't bother creating a texture if we're not going
+                    // to render anything: emplace a null value as a marker
+                    // for the regular translucent render pass.
+                    this._prerenderedTextures[layer.id] = null;
+                } else {
+                    if (layer.source !== (sourceCache && sourceCache.id)) {
+                        sourceCache = this.style.sourceCaches[layer.source];
+                        coords = [];
+
+                        if (sourceCache) {
+                            if (sourceCache.prepare) sourceCache.prepare();
+                            this.clearStencil();
+                            coords = sourceCache.getVisibleCoordinates();
+                        }
+
+                        coords.reverse();
+                    }
+
+                    if (!coords.length) {
+                        this._prerenderedTextures[layer.id] = null;
+                        continue;
+                    }
+
+                    if (!fboAttached) {
+                        // This is the first 3D layer we're rendering: attach
+                        // the framebuffer now.
+                        this._setup3DFramebuffer();
+                        // Wait to flip the boolean until after we attach the first
+                        // texture and clear the depth buffer a few lines down.
+                    }
+
+                    let renderTarget = this.viewportTextures.pop();
```

Comment:
It looks like `this.viewportTextures` will always be an empty array. Am I missing somewhere where textures get released back (say when the layer is removed from the style)?

Instead of having a `this._prerenderedTextures` map keyed on layer ID, could fill extrusion layers have a `prerenderedTexture` property?

Reply:
> Am I missing somewhere where textures get released back (say when the layer is removed from the style)?

Ah, I forgot this — added in https://github.com/mapbox/mapbox-gl-js/pull/5101/commits/0651dcfa32fec3bd4d453f5cbf8008ff25ddf1ac. 

In order for fill extrusion layers to have a `prerenderedTexture` property, we'd have to make fbos/textures members of style layers, right? Which feels weird…